src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java

```diff
@@ -118,6 +118,21 @@ public InputSource resolveEntity(String publicId, String systemId) throws SAXExc
             NodeList classes = entry.getKey();
             List<String> sourceDirs = entry.getValue();
 
+            // Collect all filenames in coverage report
+            List<String> fileNames = new ArrayList<String>();
+            for (int i = 0; i < classes.getLength(); i++) {
+                Node classNode = classes.item(i);
+                String fileName = classNode.getAttributes().getNamedItem(NODE_FILENAME).getTextContent();
+
+                if (includeFileNames != null && !includeFileNames.contains(FilenameUtils.getName(fileName))) {
+                    continue;
+                }
+                fileNames.add(fileName);
+            }
+
+            // Make multiple guesses on which of the `sourceDirs` contains files in question
+            Map<String, String> detectedSourceRoots = new PathResolver(workspace, sourceDirs).choose(fileNames);
+
             // Loop over all files in the coverage report
             for (int i = 0; i < classes.getLength(); i++) {
```

Comment:
can replace this for loop with `for(fileName: fileNames)`

Reply:
We need xml nodes:
```
            // Loop over all files in the coverage report
            for (int i = 0; i < classes.getLength(); i++) {
                Node classNode = classes.item(i);
                String fileName = classNode.getAttributes().getNamedItem(NODE_FILENAME).getTextContent();

                if (includeFileNames != null && !includeFileNames.contains(FilenameUtils.getName(fileName))) {
                    continue;
                }

                fileName = join(detectedSourceRoots.get(fileName), fileName);

                SortedMap<Integer, Integer> hitCounts = internalCounts.get(fileName);
                if (hitCounts == null) {
                    hitCounts = new TreeMap<Integer, Integer>();
                }

                NodeList children = classNode.getChildNodes();
```
Can you be specific?

Reply:
im saying we are looping twice over the same things. There is opportunity to DRY something.

SInce we already know filenames we care about from previous loop, we dont have to repeat the same logic again